W-12230783 document openshift routes (#935)
mulesoft/docs-runtime-fabric: modules/ROOT/pages/configure-openshift-routes.adoc

= Configuring Red Hat OpenShift Routes in Runtime Fabric

Runtime Fabric enables you to configure Red Hat OpenShift Routes to create ingress endpoints for your Mule applications. 

== Before You Begin

Configuration of routes in Runtime Fabric follows closely to ingress resource configuration. Before you create routes, review the following:

> * xref:custom-ingress-configuration.adoc#how-ingress-resources-templates-work-in-runtime-fabric[How Ingress Resources Templates Work in Runtime Fabric] 

Reviewer: I think you need an empty line here to make the bullet list render properly. In the GH preview it doesn't show as a bullet list, please check in local build.

* xref:custom-ingress-configuration.adoc#example-ingress-resource-templates[Example Ingress Resource Templates]

> This documentation also assumes you’re familiar with https://docs.openshift.com/container-platform/4.10/networking/routes/route-configuration.html#nw-ingress-creating-a-route-via-an-ingress_route-configuration[Creating a route through an Ingress object^].

Reviewer: Would it be better if we just add this item in the bullet list as another requisite? That way it looks more consistent and that topic seems to be a prerequisite as well.

Author: Yeah, good call!

== Example Ingress Resource Template for A Red Hat OpenShift Route

> The following example illustrates how to modify a Kubernetes ingress spec to create a Runtime Fabric ingress resource template for the networking.k8s.io/v1 API. 

Reviewer: Not sure, but I think you also need an empty line after headers. Please check in local build. 
If networking.k8s.io/v1 renders as a link in the build, you may want to prevent it by surrounding it with **+** signs:  +networking.k8s.io/v1+

[source,yaml]
--
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: example-rtf-ingress
  namespace: rtf (1)
  labels:
spec:
  ingressClassName: rtf-openshift (2)

>   Tls: (3)

Reviewer: is this correct? most values are lowercase so this could be **tls**. Not sure if values are case-sensitive but please confirm.

Author: good catch. typo, thanks!

  - hosts:
      - app-name.example.com
    secretName: example-tls
  rules:
    - host: app-name.example.com (4)
      http:
        Paths: (5)
        - path: / (6)
          pathType:
          backend: (7)
            serviceName: service-name
            servicePort: 80
--

Note the following about this example:

<1> The template must be placed in the `rtf` namespace.

> <2> `ingressClassName` must be prefixed with `rtf`-, and must be set to `rtf-openshift`.

Reviewer: ```suggestion
<2> `ingressClassName` is prefixed with `rtf-`. In this case, it is set to `rtf-openshift`.
```
I think "must be set to `rtf-openshift`" gives the idea that it should always be that value, when actually it is like that for this example to use Red Hat OpenShift right?

Author: No, in this case it has to be rtf-openshift

+

> Runtime Fabric uses the` rtf-` prefix to recognize the object as a template. Templates with `rtf-` prefix in the `spec.ingressClassName` field or in the `kubernetes.io/ingress.class annotation are consumed by the Runtime Fabric agent only and not by the actual ingress controller. Ingress controllers discover only those resources with a `spec.ingressClassName` value that use the vendor-specific name, for example,`openshift`.

Reviewer: ```suggestion
Runtime Fabric uses the `rtf-` prefix to recognize the object as a template. Templates with `rtf-` prefix in the `spec.ingressClassName` field or in the `kubernetes.io/ingress.class` annotation are consumed by the Runtime Fabric agent only and not by the actual ingress controller. Ingress controllers discover only those resources with a `spec.ingressClassName` value that use the vendor-specific name, for example,`openshift`.
```

> <3> TLS is optional. If specified in the template it will create a route with edge termination by default unless you set the `route.openshift.io/termination:` annotation.

Reviewer: ```suggestion
<3> TLS is optional. If specified in the template, Runtime Fabric creates a route with edge termination by default unless you set the `route.openshift.io/termination:` annotation.
```

> <4> The `app-name` placeholder parameter is replaced with the actual app name when you deploy the application. It helps to ensure that each endpoint name is unique if there is not a wildcard in the subdomain.

Reviewer: ```suggestion
<4> Runtime Fabric replaces the `app-name` placeholder parameter with the actual app name when you deploy the application. This helps to ensure that each endpoint name is unique if there is not a wildcard in the subdomain.
```

<5> A template can include multiple paths for a host, but Runtime Manager displays only the first path rule for the host.

> <6> The `path` parameter value is replaced by the value you add in the Path field when you configure ingress for a Mule application. Note that `path` is not a placeholder.

Reviewer: ```suggestion
<6> Runtime Fabric replaces the `path` parameter value with the value you add in the Path field when you configure ingress for a Mule application. Note that `path` is not a placeholder.
```

<7> These placeholder values are required for Kubernetes validation, but the actual values are not used by Runtime Fabric.

For a list of optional placeholder values, see xref:custom-ingress-configuration.adoc#template-placeholders[Template Placeholders].  

== Configure A Route in Runtime Fabric

To configure a route, complete the following tasks: 

. Create an ingress resource template and apply it in the cluster.
. Configure ingress for a Mule application.

=== Create an Ingress Resource Template and Apply it in the Cluster

Use the xref:example-ingress-resource-template-for-a-red-hat-openshift-route[example template] to create an ingress resource for a route.

include::partial$include-ingress-tasks.adoc[tags=createTemplate]

=== Configure Ingress for a Mule Application

include::partial$include-ingress-tasks.adoc[tags=configureIngressRuntimeManager]

> After you deploy the application, a route will be created to serve traffic to the application as per the template specified.

Reviewer: ```suggestion
After you deploy the application, Runtime Manager creates a route to serve traffic to the application as per the template specified.
```

== Configure Passthrough TLS Termination

You can configure the ingress resource template to include passthrough TLS termination. 

For more information on configuring routes with passthrough TLS termination, refer to the https://docs.openshift.com/container-platform/4.10/networking/routes/route-configuration.html#nw-ingress-creating-a-route-via-an-ingress_route-configuration[Red Hat OpenShift documentation^].

> . Follow the steps to create an ingress resource template. 

Reviewer: ```suggestion
Follow these steps to create an ingress resource template:
```
